File: Szeminarium_2020_tavasz/Gyarmati_Sandor/marcius/hirvivo.py
from sys import stdin, stdout
sets = []
counters = []
darab = 5
def root(a):
    #print(a)
    #global sets
    #print(sets)
    if sets[a] == a:
        return a
    sets[a] = root(sets[a])
    return sets[a]
def unions(a, b):
    #global sets
    #global counters
    global darab
    a = root(a)
    b = root(b)
    if a == b:
        return
    darab= darab-1
    if counters[a] >= counters[b]:
        counters[a] += counters[b]
        sets[b] = a
    else:
        counters[b] += counters[a]
        sets[a] = b
def main():
    N, M, H, K = list(map(int, stdin.readline().split()))
    #print(N, M, H, K)
    #print([] * 2)
    pontok = [0] * H
    #print(pontok)
    elek = [0] * (H * H)
    #print(elek)
    elct = 0
    for i in range(H):
        a = stdin.readline().split()
        #print(a)
        
        pontok[i] = (int(a[0]), int(a[1]))
        for j in range(i):
            hossz = abs(pontok[i][0] - pontok[j][0]) + abs(pontok[i][1] - pontok[j][1]) 
            elek[elct] = ( i, j, hossz )
            elct = elct+1
    elek = elek[:elct]
    elek.sort(key=lambda x: x[2])
    #for a in elek:
        #print(elek)
    global counters
    global darab
    global sets
    darab = H
    counters = [1] * H
    
    sets = list(range(H))
    #print(sets)
    i = 0
    L = 0
    while(i < elct and darab > K):
        L = elek[i][2]
        unions(elek[i][0], elek[i][1])
        i += 1
    print(L)
main()

File: Szeminarium_2020_tavasz/Gyarmati_Sandor/marcius/hirvivo.cpp
#include <iostream>
#include <vector>
#include <algorithm>

#define ALL(x) x.begin(), x.end()
using namespace std;

struct el{
    int aInd, bInd, hossz;
};

vector<int> sets;
vector<int> counters;
int darab;

int root(int a){
    if (sets[a] == a){
        return a;
    }
    return sets[a] = root(sets[a]);
}

void unions(int a, int b){
    a = root(a);
    b = root(b);
    if (a == b){
        return;
    }
    darab--;
    //cerr << a << " and " << b << " is now unioned\n";
    if(counters[a] >= counters[b] ){
        counters[a] += counters[b];
        sets[b] = a;
    }else{
        counters[b] += counters[a];
        sets[a] = b;
    }
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    int N, M, H, K;
    cin >> N >> M >> H >> K;
    vector< pair<int, int> > p(H);
    vector< el > elek;
    for(int i = 0; i < H; i++){
        cin >> p[i].first >> p[i].second;
        for(int j = 0; j < i; j++ ){
            int hossz = ( abs(p[i].first - p[j].first) + abs(p[i].second - p[j].second)  );
            elek.push_back( {i, j, hossz} );
            //elek.push_back( {i, j, ( abs(p[i].first - p[j].first) + abs(p[i].second - p[j].second)  )} );
        }
    }
    sort(elek.begin(), elek.end(), [](const el &a, const el &b){return a.hossz < b.hossz;});
    
    /*
    cerr << "Az elek novekvo sorrendben: \n";
    for(int i = 0; i < elek.size(); i++){
        cerr << elek[i].aInd << " " << elek[i].bInd << " : " << elek[i].hossz << endl;
    }
    */

    counters.resize(H, 1);
    sets.resize(H);
    darab = H;
    int L;
    for(int i = 0; i < H; i++){
        sets[i] = i;
    }
    for(int i = 0; i < elek.size() && darab > K; i++){
        L = elek[i].hossz;
        unions(elek[i].aInd, elek[i].bInd);
    }
    cout << L << endl;
}